src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/DateOnlyConverter.cs

```diff
@@ -47,12 +47,14 @@ private static DateOnly ReadCore(ref Utf8JsonReader reader)
                 source = stackSpan.Slice(0, bytesWritten);
             }
 
-            if (!JsonHelpers.TryParseAsIso(source, out DateOnly value))
+            if (JsonHelpers.IsInRangeInclusive(source.Length, FormatLength, MaxEscapedFormatLength) &&
```

Comment:
Why is this additional check necessary? I assumed the same check in line 33 does take escaping into account.

Reply:
Can we move it into the `else` after the copy?

Reply:
> Line 46: `int bytesWritten = reader.CopyString(stackSpan);` performs unescaping which can change the length of the source.

I guess I'm confused by the fact that this check keeps using `MaxEscapedFormatLength` as the upper bound even though `source` is guaranteed to be unescaped at this point. 

Judging by the test you've included a lower bound check would suffice, but I suspect the problem really lies in the fact that the `TryParseAsIso` method isn't handling short inputs correctly (it's throwing presumably?)

Reply:
>TryParseAsIso method isn't handling short inputs correctly (it's throwing presumably?)

It has a `Debug.Assert`
https://source.dot.net/#System.Text.Json/System/Text/Json/JsonHelpers.Date.cs,146
I chose not to change it to a runtime check in order to reduce noise from changing all the callsites that _do_ properly check the length.

It then starts accessing various positions in the `Span` assuming there are at least 10 digits available, so at runtime you could get `ArgumentException`.

> Judging by the test you've included a lower bound check would suffice

Sure, I can change it to `if (bytesWritten < FormatLength)`